### ISSUE TITLE
API mode

### ISSUE THREAD

**Issue by qr7hur** [MaximeGILLET/Quotes#6]
Having an API (or "API mode") could allow people to push front-end enhancements and suggest API endpoints to back-end devs without actually having to setup an IDE :)

Plus, many "fashionable" frameworks would benefit from having a REST API (React, Angular, Polymer ...etc).

**Comment by Nerevarin117** [comment]
Maybe interesting for nkt++ project but what would be the addition and the interest for MyQuotes?

"people" you mean any anonymous random guy ending on the website or clearly identified and trusted users?

**Comment by qr7hur** [comment]
Eg1:

Hey in order to have a live feed, I need an endpoint which provides me with a filtered list of quotes matching my user's preference. Then I can XMLHttpRequest it every 10 seconds for seamless updates.
- Endpoint: GET /feed
- Parameters: UserID
- Returns for example a JSON with response metadata (time...) and an array of {user, quote, time} for the feed.

Eg2:

Hey in order to design the LIKE button, I need two endpoints:
- Endpoint GET: /like/:quoteID
- Parameters: UserID
- Returns for example a JSON with response metadata (time...) and a number_of_likes: XXX entry.
- Endpoint POST: /like/:quoteID
- Parameters: UserID
- Returns informational JSON about the success/failure of the action...

So the front-end can asynchronously request the back-end for specific elements on the page...

**Comment by qr7hur** [comment]
By "people" I meant me.

**Comment by Nerevarin117** [comment]
Such an api can be implemented if other/future applications feel the need to read quote data, BUT it will sure be readonly. Basically  every method of our controllers will return json responses, so the functionnality might be there already. Maybe we will provide later on the dedicated url not to break the api if any evolution of the model occures and to keep tracks on the version.

**Comment by Nerevarin117** [comment]
If we do so it will be a separated project I think.

**Comment by qr7hur** [comment]
Dude I'm talking about a client API for our own use: our front-end would be talking with our back-end through the API. I'm not talking about opening it to third parties.

**Comment by Nerevarin117** [comment]
Well you can already do that as most of our controllers return json objects, we just need some setup about the routePath and the json formats.

If you think of your backdoor with the javascript/html comming from another source that is an obvious NO.

**Comment by qr7hur** [comment]
Yeah obviously that's not what I'm talking about. I'm just splitting the monolithic application into two parts which are front end and back end.
What I imagine is the following:
- The back-end (server-side) deals with extracting and modeling the data, serving it to the client.
- The front-end (client-side) deals with rendering the data into a view (maybe we can find a way to pre-render on the back-end for perf.), and providing feature-related logic, then sends data back to the server on user input.
- The back-end (server-side) receives the data, makes necessary coherence- and security- related logic, then saves it according to its model.

This way maybe we can more easily share tasks between back/front (since I can't test the back-end right now...). Also it's flexible...etc

**Comment by Nerevarin117** [comment]
MVC does already no?
